crypto/ec/ec_lib.c

```diff
@@ -1012,11 +1012,59 @@ int ec_group_simple_order_bits(const EC_GROUP *group)
     return BN_num_bits(group->order);
 }
 
+static int ec_field_inverse_mod_ord(const EC_GROUP *group, BIGNUM *r,
+                                    BIGNUM *x, BN_CTX *ctx)
+{
+    BIGNUM *e = NULL;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
+        goto err;
+
+    BN_CTX_start(ctx);
+    if ((e = BN_CTX_get(ctx)) == NULL)
+        goto err;
+
+    /* Check if optimized inverse is implemented */
+    if (group->mont_data != NULL) {
+        /*-
+         * We want inverse in constant time, therefore we utilize the fact
+         * order must be prime and use Fermats Little Theorem instead.
+         */
+        if (!BN_set_word(e, 2))
+            goto err;
+        if (!BN_sub(e, group->order, e))
+            goto err;
+        /*-
+         * Exponent e is public.
+         * No need for scatter-gather or BN_FLG_CONSTTIME.
+         */
+        if (!BN_mod_exp_mont(r, x, e, group->order, ctx, group->mont_data))
+            goto err;
+        /* Inverse of zero doesn't exist. Let the fallback catch it. */
+        ret = (BN_is_zero(r)) ? 0 : 1;
```

Comment:
Why run the fallback? It's already known there's no inverse.

Reply:
To have it throw a consistent error with the `group->order` even case, to catch in unit testing. But per your API suggestion, it sounds like this will not be able to be directly unit tested anyway, so I can strip that out as soon as I get a clear consensus from the OpenSSL team on what `EC_GROUP_do_inverse_ord` is supposed to do. (Alternatively, I can be bold and just guess / document.)